### PR TITLE
Fix progress bars getting stuck at 0%

### PR DESCRIPTION
**Pull request by FlorentinD**

### CODE REVIEW

File: graphdatascience/query_runner/neo4j_query_runner.py [hunk 1]
```diff
@@ -248,17 +248,20 @@ def _log(self, job_id: str, future: "Future[Any]", database: Optional[str] = Non
...
 
             progress_percent = progress["progress"][0]
-            if not progress_percent == "n/a":
-                task_name = progress["taskName"][0].split("|--")[-1][1:]
-                pbar = pbar or tqdm(total=100, unit="%", desc=task_name)
-            else:
+            if progress_percent == "n/a":
                 return
 
+            task_name = progress["taskName"][0].split("|--")[-1][1:]
```
Comment (FlorentinD) [inline graphdatascience/query_runner/neo4j_query_runner.py:254]: I dont understand why we only take the first taskName and not update all related bars

File: graphdatascience/query_runner/neo4j_query_runner.py [hunk 2]
```diff
@@ -248,17 +248,20 @@ def _log(self, job_id: str, future: "Future[Any]", database: Optional[str] = Non
...
+                pbars[task_name] = tqdm(total=100, unit="%", desc=task_name)
+            pbar = pbars[task_name]
+
             parsed_progress = float(progress_percent[:-1])
             pbar.update(parsed_progress - pbar.n)
 
-        if pbar:
+        for pbar in pbars.values():
             pbar.update(100 - pbar.n)
+            pbar.refresh()
```
Comment (FlorentinD) [inline graphdatascience/query_runner/neo4j_query_runner.py:264]: this is the actual fix

**Comment by FlorentinD** [comment]
[Empty progress bars](https://trello.com/c/ZO4alwzt/9343-empty-progress-bars)

File: graphdatascience/query_runner/neo4j_query_runner.py (hunk 2 above)
Comment (adamnsch) [reply graphdatascience/query_runner/neo4j_query_runner.py:264]: (and the one in send_df 😉 )

File: graphdatascience/query_runner/neo4j_query_runner.py [hunk 3]
```diff
@@ -248,17 +248,20 @@ def _log(self, job_id: str, future: "Future[Any]", database: Optional[str] = Non
...
-                task_name = progress["taskName"][0].split("|--")[-1][1:]
-                pbar = pbar or tqdm(total=100, unit="%", desc=task_name)
-            else:
+            if progress_percent == "n/a":
                 return
 
+            task_name = progress["taskName"][0].split("|--")[-1][1:]
+            if task_name not in pbars:
+                pbars[task_name] = tqdm(total=100, unit="%", desc=task_name)
+            pbar = pbars[task_name]
```
Comment (adamnsch) [inline graphdatascience/query_runner/neo4j_query_runner.py:257]: Why do we need this dictionary? Don't we only have one progress bar?

**Review by adamnsch — COMMENTED** [review]
Nice work! I have a question

File: graphdatascience/query_runner/neo4j_query_runner.py (hunk 3 above)
Comment (FlorentinD) [reply graphdatascience/query_runner/neo4j_query_runner.py]: I thought its needed as my first assumption was result ordering, but in the end its not as we only retrieve the root task.

I added a `LIMIT 1` in the cypher query to make this clearer.

File: changelog.md [hunk 4]
```diff
@@ -20,6 +20,7 @@
 * Fixed an issue where source and target IDs of relationships in heterogeneous OGBL graphs were not parsed correctly.
 * Fixed an issue where configuration parameters such as `aggregation` were ignored by `gds.graph.toUndirected`.
 * Fixed an issue where the `database` given for the `GraphDataScience` construction was not used for metadata retrieval, causing an exception to be raised if the default "neo4j" database was missing.
+* Fixed an issue where progress bars would not complete.
```
Comment (adamnsch) [inline changelog.md:23]:
```suggestion
* Fixed an issue where progress bars would not always complete.
```
As I think it worked most of the time

**Review by adamnsch — APPROVED** [review]
Very nice 👏

**Pull request merged**